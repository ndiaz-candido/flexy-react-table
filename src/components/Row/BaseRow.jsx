import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  LeftSection,
  RightSection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width: `${width}px`,
  flex: `0 0 ${width}px`,
});

class BaseRow extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({ columns, rowData, columnClassName }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const columnWidth = column.width || BaseRow.DEFAULT_COLUMN_WIDTH;
      const columnDataKey = column.dataKey || '';
      const cellContent = rowData[columnDataKey] || '';
      const styles = cellDefaultStyles(columnWidth);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (column.stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { className, isSticky, onScroll, rowId, scrollLeft } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftSection cells={leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightSection cells={rightStickySection} />
      </div>
    );
  }
}

BaseRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
  columnClassName: PropTypes.any,
  isSticky: PropTypes.bool,
  isHeader: PropTypes.bool,
  onScroll: PropTypes.func,
  rowId: PropTypes.number,
  rowData: PropTypes.object,
  scrollLeft: PropTypes.number,
};

BaseRow.defaultProps = {
  columns: [],
  className: '',
  columnClassName: '',
  isSticky: false,
  isHeader: false,
  onScroll: () => {},
  rowId: -1,
  rowData: {},
  scrollLeft: 0,
};

export default BaseRow;
