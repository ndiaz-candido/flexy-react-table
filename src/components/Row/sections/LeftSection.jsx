import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

/* eslint-disable react/no-array-index-key */

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  left: '0',
};
const LeftSection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    {cells.map((cell, index) => (
      <Fragment key={`leftSection-${index}`}>{cell}</Fragment>
    ))}
  </div>
);

LeftSection.propTypes = {
  cells: PropTypes.array,
};

LeftSection.defaultProps = {
  cells: [],
};

export default LeftSection;
