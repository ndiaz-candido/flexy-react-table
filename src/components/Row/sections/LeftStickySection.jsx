import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  left: '0',
};
const LeftStickySection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    { cells.map((cell, index) => <Fragment key={`leftSection-${index}`}>{cell}</Fragment>) }
  </div>
);

LeftStickySection.propTypes = {
  cells: PropTypes.array,
};

LeftStickySection.defaultProps = {
  cells: []
};

export default LeftStickySection;
