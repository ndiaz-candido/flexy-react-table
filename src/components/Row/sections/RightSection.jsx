import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

/* eslint-disable react/no-array-index-key */

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  right: '0',
};
const RightStickySection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    {cells.map((cell, index) => (
      <Fragment key={`rightSection=${index}`}>{cell}</Fragment>
    ))}
  </div>
);

RightStickySection.propTypes = {
  cells: PropTypes.array.isRequired,
};

export default RightStickySection;
