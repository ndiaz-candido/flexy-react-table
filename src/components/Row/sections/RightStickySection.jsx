import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const defaultSectionStyle = {
  display: 'flex',
  flex: '0 0 auto',
  position: 'sticky',
  right: '0',
};
const RightStickySection = ({ cells }) => (
  <div style={defaultSectionStyle}>
    {cells.map((cell, index) => (<Fragment key={`rightSection-${index}`}>{cell}</Fragment>))}
  </div>
);

RightStickySection.propTypes = {
  cells: PropTypes.array,
};

RightStickySection.defaultProps = {
  cells: []
};

export default RightStickySection;
