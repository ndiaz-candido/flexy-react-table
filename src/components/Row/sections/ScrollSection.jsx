import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Scrollbars } from 'react-custom-scrollbars';
import idGenerator from 'react-id-generator';

/* eslint-disable react/no-array-index-key */
/* eslint-disable react/no-multi-comp */

const defaultWrapperStyle = {
  flex: '1 1 auto',
};
const defaultSectionStyle = {
  display: 'flex',
};

/**
 * This PureComponent avoids updating the cells when the user
 * is scrolling improving the performance.
 */
class Cell extends PureComponent {
  render() {
    return this.props.children;
  }
}

class ScrollSection extends PureComponent {
  state = {
    showScrollTrack: false,
  };

  handleOnMouseEnter = () => this.setState({ showScrollTrack: true });

  handleOnMouseLeave = () => this.setState({ showScrollTrack: false });

  handleOnScroll = () =>
    this.props.onScroll({ left: this.scrollableArea.getScrollLeft() });

  renderScrollView = ({ style, ...props }) => {
    const { showScrollTrack } = this.state;
    const showXScrollTrack = { overflowX: 'scroll' };
    const hideXScrollTrack = { overflowX: 'hidden' };
    const customStyles = showScrollTrack ? showXScrollTrack : hideXScrollTrack;

    return (
      <div
        key={idGenerator()}
        style={{
          ...style,
          ...defaultSectionStyle,
          ...customStyles,
        }}
        {...props}
      />
    );
  };

  setScrollableAreaRef = element => {
    const { rowId, registerScrollSection } = this.props;

    this.scrollableArea = element;

    registerScrollSection(rowId, this.scrollableArea);
  };

  render() {
    const { cells } = this.props;

    return (
      <div
        id="scrollable-wrapper"
        style={defaultWrapperStyle}
        onMouseEnter={this.handleOnMouseEnter}
        onMouseLeave={this.handleOnMouseLeave}>
        <Scrollbars
          ref={this.setScrollableAreaRef}
          renderView={this.renderScrollView}
          onScroll={this.handleOnScroll}>
          {cells.map((cell, index) => (
            <Cell key={`scroll-${index}`}>{cell}</Cell>
          ))}
        </Scrollbars>
      </div>
    );
  }
}

ScrollSection.propTypes = {
  cells: PropTypes.array,
  onScroll: PropTypes.func,
  registerScrollSection: PropTypes.func,
};

ScrollSection.defaultProps = {
  cells: [],
  onScroll: () => {},
  registerScrollSection: () => {},
};

export default ScrollSection;
