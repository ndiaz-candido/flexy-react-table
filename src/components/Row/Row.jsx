import React, { Component } from 'react';
import PropTypes from 'prop-types';
import idGenerator from 'react-id-generator';
import { ScrollSection, RightSection, LeftSection } from './sections';
import ScrollContext from '../ScrollContext';

const defaultRowStyle = {
  display: 'flex',
  minHeight: '50px',
};
const defaultCellStyle = width => ({
  width,
  flex: `0 0 ${width}`,
});

class Row extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;

  state = {
    stickRightSection: [],
    stickLeftSection: [],
    scrollSection: [],
  };

  constructor(props) {
    super(props);

    this.rowId = idGenerator();
  }

  static getDerivedStateFromProps({ columns, cellClassName, rowData }) {
    const stickLeftSection = [];
    const stickRightSection = [];
    const scrollSection = [];

    columns.forEach(column => {
      const columnDataKey = column.dataKey;
      const columnWidth = column.width || Row.DEFAULT_COLUMN_WIDTH;
      const cellContent = rowData[columnDataKey];
      const cell = (
        <div className={cellClassName} style={defaultCellStyle(columnWidth)}>
          {cellContent}
        </div>
      );

      switch (column.stickAlign) {
        case 'right':
          stickRightSection.push(cell);
          break;
        case 'left':
          stickLeftSection.push(cell);
          break;
        default:
          scrollSection.push(cell);
          break;
      }
    });

    return {
      stickLeftSection,
      stickRightSection,
      scrollSection,
    };
  }

  render() {
    const { stickLeftSection, stickRightSection, scrollSection } = this.state;
    const { className } = this.props;

    return (
      <div className={className} style={defaultRowStyle}>
        <LeftSection cells={stickLeftSection} />
        <ScrollContext.Consumer>
          {({ handleOnScroll, registerScrollSection }) => (
            <ScrollSection
              cells={scrollSection}
              onScroll={handleOnScroll}
              registerScrollSection={registerScrollSection}
              rowId={this.rowId}
            />
          )}
        </ScrollContext.Consumer>
        <RightSection cells={stickRightSection} />
      </div>
    );
  }
}

Row.propTypes = {
  className: PropTypes.any,
  cellClassName: PropTypes.any,
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      dataKey: PropTypes.string,
      stickAlign: PropTypes.string,
      width: PropTypes.string,
    })
  ),
  rowData: PropTypes.object,
};

Row.defaultProps = {
  className: '',
  cellClassName: '',
  columns: [],
  rowData: {},
};

export default Row;
