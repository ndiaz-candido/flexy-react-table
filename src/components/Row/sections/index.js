import LeftSection from './LeftSection';
import RightSection from './RightSection';
import ScrollableSection from './ScrollableSection';

export { LeftSection, RightSection, ScrollableSection };
