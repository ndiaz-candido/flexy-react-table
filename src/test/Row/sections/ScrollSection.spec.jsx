import React from 'react';
import { shallow, mount } from 'enzyme';
import { ScrollSection } from '../../../components/Row/sections';

describe('ScrollSection', () => {
  const cells = [
    <div key="0">Cell A</div>,
    <div key="1">Cell B</div>,
    <div key="2">Cell C</div>,
  ];

  test('should render its default content', () => {
    const ScrollSectionComponent = shallow(<ScrollSection cells={[]} />);

    expect(ScrollSectionComponent).toMatchSnapshot();
  });

  test('should render the given cells', () => {
    const ScrollSectionComponent = shallow(<ScrollSection cells={cells} />);

    expect(ScrollSectionComponent).toMatchSnapshot();
  });

  test('should render the scroll arrows when the section is scrollable', () => {
    const ScrollSectionComponent = shallow(<ScrollSection cells={cells} />);

    ScrollSectionComponent.setState({ shouldShowScrollArrows: true });

    expect(ScrollSectionComponent).toMatchSnapshot();
  });

  test('should call the callback when the section is scrolled to the left', () => {
    const onScroll = jest.fn();
    const ScrollSectionComponent = mount(
      <ScrollSection cells={cells} onScroll={onScroll} />
    );

    ScrollSectionComponent.setState({ shouldShowScrollArrows: true });

    const ScrollBars = ScrollSectionComponent.find('#scrollLeft');

    ScrollBars.prop('onClick')();

    expect(onScroll).toHaveBeenCalledWith(
      expect.objectContaining({ direction: 'left' })
    );
  });

  test('should call the callback when the section is scrolled to the right', () => {
    const onScroll = jest.fn();
    const ScrollSectionComponent = mount(
      <ScrollSection cells={cells} onScroll={onScroll} />
    );

    ScrollSectionComponent.setState({ shouldShowScrollArrows: true });

    const ScrollBars = ScrollSectionComponent.find('#scrollRight');

    ScrollBars.prop('onClick')();

    expect(onScroll).toHaveBeenCalledWith(
      expect.objectContaining({ direction: 'right' })
    );
  });

  test('should register the scroll section when it renders', () => {
    const registerScrollSectionElements = jest.fn();

    mount(
      <ScrollSection
        cells={[]}
        registerScrollSectionElements={registerScrollSectionElements}
      />
    );

    expect(registerScrollSectionElements).toHaveBeenCalled();
  });
});
