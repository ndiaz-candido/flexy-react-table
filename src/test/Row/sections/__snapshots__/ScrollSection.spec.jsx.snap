// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollSection should render its default content 1`] = `
<div
  id="scrollable-wrapper"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "flex": "1 1 auto",
      "position": "relative",
    }
  }
>
  <div
    style={
      Object {
        "backgroundImage": "linear-gradient(to right, rgba(202, 203, 204, 0), rgba(202, 203, 204, 1))",
        "height": "100%",
        "position": "absolute",
        "right": 0,
        "width": "20px",
      }
    }
  />
  <div
    style={
      Object {
        "backgroundImage": "linear-gradient(to left, rgba(202, 203, 204, 0), rgba(202, 203, 204, 1))",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "width": "20px",
      }
    }
  />
  <Scrollbars
    autoHeight={false}
    autoHeightMax={200}
    autoHeightMin={0}
    autoHide={false}
    autoHideDuration={200}
    autoHideTimeout={1000}
    hideTracksWhenNotNeeded={false}
    onScroll={[Function]}
    renderThumbHorizontal={[Function]}
    renderThumbVertical={[Function]}
    renderTrackHorizontal={[Function]}
    renderTrackVertical={[Function]}
    renderView={[Function]}
    tagName="div"
    thumbMinSize={30}
    universal={false}
  />
</div>
`;

exports[`ScrollSection should render the given cells 1`] = `
<div
  id="scrollable-wrapper"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "flex": "1 1 auto",
      "position": "relative",
    }
  }
>
  <div
    style={
      Object {
        "backgroundImage": "linear-gradient(to right, rgba(202, 203, 204, 0), rgba(202, 203, 204, 1))",
        "height": "100%",
        "position": "absolute",
        "right": 0,
        "width": "20px",
      }
    }
  />
  <div
    style={
      Object {
        "backgroundImage": "linear-gradient(to left, rgba(202, 203, 204, 0), rgba(202, 203, 204, 1))",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "width": "20px",
      }
    }
  />
  <Scrollbars
    autoHeight={false}
    autoHeightMax={200}
    autoHeightMin={0}
    autoHide={false}
    autoHideDuration={200}
    autoHideTimeout={1000}
    hideTracksWhenNotNeeded={false}
    onScroll={[Function]}
    renderThumbHorizontal={[Function]}
    renderThumbVertical={[Function]}
    renderTrackHorizontal={[Function]}
    renderTrackVertical={[Function]}
    renderView={[Function]}
    tagName="div"
    thumbMinSize={30}
    universal={false}
  >
    <div
      key="0"
    >
      Cell A
    </div>
    <div
      key="1"
    >
      Cell B
    </div>
    <div
      key="2"
    >
      Cell C
    </div>
  </Scrollbars>
</div>
`;
